asm/sjis.s

```diff
@@ -0,0 +1,206 @@
+BEGIN Kanji_OffsetFromShiftJIS
+    # Characters with codepoints >= 0x8800 are kanji. Arrangement is regular, 
+    # so convert index directly.
+    li      $at, 0x8800
+    slt     $at, $a0, $at
+    bnez    $at, .not_kanji
+     # 0xBC is the size of one block of glyphs in `kanji`:
+     # 0x100 possible codepoints with the same byte1
+     # - 0x40 unused at beginning 
+     # - 1 unused at 0x7F 
+     # - 3 unused at 0xFD, 0xFE, 0xFF
+     li     $a2, 0xBC
+    # Get byte1 and adjust so starts at 0
+    srl     $a1, $a0, 8
+    addi    $a1, $a1, -0x88
+    multu   $a2, $a1
+    # Get byte2 and adjust so starts at 0
+    andi    $a3, $a0, 0xff
+    addi    $a3, $a3, -0x40
+    slti    $at, $a3, 0x40
+    mflo    $a2
+    # 0xXX7F is always empty and elided in the file, so decrement
+    bnezl   $at, .kanji_lower_block
+     mflo   $a2
+    addi    $a3, $a3, -1
+    mflo    $a2
+.kanji_lower_block:
+    addi    $a3, $a3, 0x30a
+    add     $a3, $a3, $a2
+    jr      $ra
+     sll    $v0, $a3, 7
+     # returns (0x30A + (adjusted byte2) + (adjusted byte1) * 0xBC) * FONT_CHAR_TEX_SIZE
+
+# Non-kanji are arranged with irregular gaps, use the lookup table.
+.not_kanji:
+    # Get byte1 and adjust so starts at 0
+    srl     $a1, $a0, 8
+    addi    $a1, $a1, -0x81
+    multu   $a2, $a1
+    # Get byte2 and adjust so starts at 0
+    andi    $a3, $a0, 0xff
+    addi    $a3, $a3, -0x40
+    slti    $at, $a3, 0x40
+    mflo    $a2
+    # 0xXX7F is always empty and elided in the file, so decrement
+    bnezl   $at, .not_kanji_lower_block
+     mflo   $a2
+    addi    $a3, $a3, -1
+    mflo    $a2
+.not_kanji_lower_block:
+    add     $a3, $a3, $a2
+    lui     $a2, %hi(kanjiNonKanjiIndices)
+    sll     $a3, $a3, 1
+    addiu   $a2, %lo(kanjiNonKanjiIndices)
+    add     $a3, $a3, $a2
+    lh      $a2, ($a3)
+    jr      $ra
+     sll    $v0, $a2, 7
+     # returns kanjiNonKanjiIndices[(adjusted byte2) + (adjusted byte1) * 0xBC] * FONT_CHAR_TEX_SIZE
+END Kanji_OffsetFromShiftJIS
+
+# C pseudocode:
+# s32 Kanji_OffsetFromShiftJIS(s32 arg0) {
+#     s32 byte1;
+#     s32 byte2;
+#
+#     if (arg0 >= 0x8800) {
+#         byte1 = arg0 >> 8;
+#         byte1 -= 0x88;
+#
+#         byte2 = (arg0 & 0xFF);
+#         byte2 -= 0x40;
+#         if (byte2 >= 0x40) {
+#             byte2--;
+#         }
+#
+#         return (0x30A + byte2 + byte1 * 0xBC) * 0x80;
+#     } else {
+#         byte1 = arg0 >> 8;
+#         byte1 -= 0x81;
+#
+#         byte2 = (arg0 & 0xFF);
+#         byte2 -= 0x40;
+#         if (byte2 >= 0x40) {
+#             byte2--;
+#         }
+#
+#         return kanjiNonKanjiIndices[byte2 + byte1 * 0xBC] * 0x80;
+#     }
+# }
+
+BEGINDATA kanjiNonKanjiIndices
+    #       ___0    ___1    ___2    ___3    ___4    ___5    ___6    ___7    ___8    ___9    ___A    ___B    ___C    ___D    ___E    ___F
+    .half 0x0000, 0x0001, 0x0002, 0x0003, 0x0004, 0x0005, 0x0006, 0x0007, 0x0008, 0x0009, 0x000A, 0x000B, 0x000C, 0x000D, 0x000E, 0x000F # 814_
```

Comment:
```suggestion
    #       __0    __1    __2    __3    __4    __5    __6    __7    __8    __9    __A    __B    __C    __D    __E   __F
    .half 0x000, 0x001, 0x002, 0x003, 0x004, 0x005, 0x006, 0x007, 0x008, 0x009, 0x00A, 0x00B, 0x00C, 0x00D, 0x00E, 0x00F # 814_
```
I think you could remove at least 1 leading zero as there isn't any number in this data that uses 4 digits. I would even suggest removing all leading 0's if the formatter doesn't break the nice alignment you've got going (or you can leave it otherwise)

Reply:
dont believe formatter is touching .s files so dont think thats a concern (someone please correct me if wrong though)

Reply:
> dont believe formatter is touching .s files so dont think thats a concern (someone please correct me if wrong though)

It shouldn't be, and I doubt it would produce anything meaningful for non-C code (would probably add a final newline).

> I think you could remove at least 1 leading zero as there isn't any number in this data that uses 4 digits. I would even suggest removing all leading 0's if the formatter doesn't break the nice alignment you've got going (or you can leave it otherwise)

I left them like this mainly because they're accessed as halfs. I've thought for a while about how to format this data, and I think this is probably the least offensive way for now. In theory one could label all of them with the symbols, although there are some curious repeats, and holes, and to do so makes the file over 1000 lines long (there are 0x524 entries in this table, I believe).